pmlc/dialect/stripe/populate_tensor_ref_shape_pass.cc

```diff
@@ -0,0 +1,90 @@
+// Copyright 2019 Intel Corporation
+
+#include <algorithm>
+
+#include "pmlc/dialect/stripe/analysis.h"
+#include "pmlc/dialect/stripe/dialect.h"
+#include "pmlc/dialect/stripe/populate_tensor_ref_shape_pass.h"
+
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#define PASS_NAME "populate-tensor-ref-shape"
+#define DEBUG_TYPE PASS_NAME
+
+using mlir::BlockArgument;
+using mlir::FuncOp;
+using mlir::FunctionType;
+using mlir::Operation;
+using mlir::SmallVector;
+using mlir::Type;
+using mlir::Value;
+using pmlc::dialect::stripe::TensorRefType;
+
+namespace {
+
+/// This pass populates the optional shape information of 'tensor_ref' types. This information is found in the 'layout'
+/// attribute of 'tensor_ref' types that are used in BlockArgument values and Alloc operations. For each 'tensor_ref'
+/// Value, the algorithm proceed as follows:
+///   1. Traverse the use-def chain until a BlockArgument/Alloc is found and retrieve its 'layout' attribute.
+///   2. Replace Value's 'tensor_ref' type with a new 'tensor_ref' type that contains the shape information.
+// NOTE: This analysis is O(n^2) so we should revisit this approach/design if the layout information is needed in
+// more cases than the Stripe->Affine dialect conversion.
+class PopulateTensorRefShape : public mlir::FunctionPass<PopulateTensorRefShape> {
```

Comment:
Wouldn't a simple pre-order walk of the program suffice?  Basically if we've 'updated' all Ops before a given op, and we've set the values on the block arguments, then each Alloc can just set it's layout directly, and each tensor op (Refine, Load, etc) can just copy it's layout (and it's whole type in fact) from it's values immediate definition (which by virtue of being earlier is certainly already updates).  This would be linear with program size.

Reply:
Thanks, Jeremy. Yes, that's a reasonable next step for the future.